feat(nfs): created a router agnostic framework

schultzp2020 opened the pull request:
Signed-off-by: Paul Schultz <pschultz@pobox.com>

## Hey, I just made a Pull Request!

This PR decouples the frontend system from a direct dependency on `react-router-dom` by introducing a router-agnostic `RouterApi` in `@backstage/frontend-plugin-api`. This abstraction defines a standard interface for routing operations—such as matching routes, generating paths, and navigation—allowing the core framework to operate independently of the underlying routing library.

The default implementation, `ReactRouter6RouterApi`, has been moved to a new package `@backstage/frontend-module-react-router-v6` and is provided via the app plugin (`@backstage/plugin-app`) as an `ApiBlueprint`. This follows the standard FrontendFeature pattern, meaning custom router implementations can be provided by supplying an alternative API factory for `routerApiRef`.

### Exported APIs

**Hooks** - Import from `@backstage/frontend-plugin-api`:
- `useLocation`, `useNavigate`, `useParams`, `useSearchParams`, `useResolvedPath`, `useHref`, `useOutlet`, `useRoutes`

**Utility functions** - Import from `@backstage/frontend-plugin-api`:
- `matchRoutes`, `generatePath`, `resolvePath`

> **Note:** These are intentionally named to match their `react-router-dom` counterparts so that migration is a simple import swap. They require component context due to internal `useApi()` lookup, which is documented in the `@remarks` JSDoc.

**Components** - Import from `@backstage/frontend-plugin-api`:
- `Link`, `NavLink`, `Outlet`, `Routes`, `Route`, `Navigate`

The `Routes` and `Route` components work around react-router's internal child type checking by using `useRoutes` internally, allowing them to be imported and used like standard React components.

### Testing

`MockMemoryRouterApi` and `MockBrowserRouterApi` are available in `@backstage/frontend-test-utils`. `MockMemoryRouterApi` uses a `MemoryRouter` internally and is suitable for most tests. `MockBrowserRouterApi` uses a `BrowserRouter` for tests that require real browser history APIs.

Convenience providers `TestMemoryRouterProvider` and `TestBrowserRouterProvider` are also available for wrapping test renders.

### React Router v7 Compatibility

A development-time warning has been added to detect multi-segment splat paths (e.g., `dashboard/*`) that have different relative link behavior in react-router v7. See: https://reactrouter.com/upgrading/v6#v7_relativesplatpath

### Usage

```tsx
import { 
  Routes, 
  Route, 
  Link, 
  useLocation, 
  useNavigate 
} from '@backstage/frontend-plugin-api';

function MyPage() {
  const location = useLocation();
  
  return (
    <Routes>
      <Route path="/" element={<Home />} />
      <Route path="/about" element={<About />} />
    </Routes>
  );
}
```

**Note:** Since the underlying implementation provided by default is still React Router v6, existing behavior remains unchanged. The primary purpose of this change is to pave the way for future migrations; once the ecosystem fully adopts the new frontend system and routing framework, we can upgrade to newer routers (such as React Router v7 or TanStack Router) without forcing widespread refactoring on adopters.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on packages/frontend-app-api/report.api.md:
```diff
@@ -158,6 +161,10 @@ export type CreateSpecializedAppOptions = {
+    router?: {
+      matchRoutes: RoutingContextType['matchRoutes'];
+      generatePath: RoutingContextType['generatePath'];
+    };
```
Do we need to expose this as something configurable? I'm thinking it's best to keep this internal to make it easier to make changes

Rugvip replied:
Point is to abstract this away though so that we can switch out the implementation in app-api without impact to consumers. Thinking at most we'd have a flag for it in these options?

Rugvip commented on packages/frontend-plugin-api/report.api.md:
```diff
@@ -1954,6 +2034,9 @@ export function useRouteRefParams<Params extends AnyRouteRefParams>(
 ): Params;
 
+// @public (undocumented)
+export function useRouting(): RoutingContextType;
```
I'm thinking the public API should most likely be the what react-router already provides, more or less? `Route`, `Routes`, `Outlet`, `useLocation`, `useNavigation`, etc. Feels like that'd be a bit of a requirement to eventually move over. With that the context type would be an internal thing. A potential alternate approach to context could be to just go with a `RoutingApi` as well, just like how we switched from the route resolution context in the old frontend system to `RouteResolutionApi` in NFS. That has already turned out to be quite useful in a couple of places.

It's a little bit scary thinking about our ability to push and verify this change though, since we can't exactly deprecate the RR methods directly. We'd need to aim for the ability to manage the full migration with lint rules and codemods I'd think.

Rugvip reviewed (commented):
Thank you for getting this ball rolling! 👍

schultzp2020 commented on packages/frontend-app-api/report.api.md:
```diff
@@ -174,4 +181,18 @@ export type FrontendPluginInfoResolver = (ctx: {
+  children,
+}: {
+  children: ReactNode;
+}) => JSX_2.Element;
```
It's exported so it can be used in `ForwardsCompatProvider.tsx` and `TestRouterProvider.tsx`.

schultzp2020 commented:
@Rugvip PTAL?

Rugvip commented on packages/frontend-plugin-api/src/blueprints/RouterBlueprint.test.tsx:
```diff
@@ -59,9 +61,9 @@ describe('RouterBlueprint', () => {
       params: {
         component: ({ children }) => (
-          <MemoryRouter>
+          <TestRouterProvider>
```
yep I mean the current API is good

schultzp2020 commented:
> @schultzp2020 replied to some things, but I think the main discussion to focus on settling before moving forward is [#31829 (comment)](https://github.com/backstage/backstage/pull/31829#discussion_r2561635715)

@Rugvip Done! Please let me know what you think. I really like the API approach you suggested. 🙂

Parsifal-M reviewed (approved):
Approving from docs side of things, makes sense to me, just a small comment but not blocking 😄 👍

schultzp2020 commented on docs/frontend-system/utility-apis/01-index.md:
```diff
@@ -43,6 +43,12 @@ When testing frontend components and extensions, you often need to provide mock
 
+## Router API
+
+> For details, [see the main article](./06-router-api.md).
```
The link target is correct — the file is `05-router-api.md`, which matches `./05-router-api.md`.

schultzp2020 commented on packages/frontend-test-utils/src/apis/RouterApi/MockBrowserRouterApi.tsx:
```diff
@@ -0,0 +1,180 @@
+  Router: ComponentType<{ children: ReactNode; basePath: string }> = ({
+    children,
+  }) => <BrowserRouter basename={this.basePath}>{children}</BrowserRouter>;
+
```
Already addressed — `MockBrowserRouterApi.Router` now accepts the `basePath` prop and passes it to `BrowserRouter` via `basename={basePath}`.

schultzp2020 committed:
feat(nfs): introduce router-agnostic framework (RouterApi)

Decouple the Backstage frontend system from react-router-dom by
introducing a RouterApi abstraction. This allows adopters to swap in
alternative router implementations without changing plugin code.

Key changes:
- Add RouterApi interface and routerApiRef in frontend-plugin-api
- Create frontend-module-react-router-v6 as the default implementation
- Add routing wrapper hooks and components in frontend-plugin-api
- Integrate RouterApi into RouteResolver, RouteTracker, and phaseApis
- Add RouterBlueprint in app-react
- Add MockMemoryRouterApi and TestMemoryRouterProvider test utilities
- Register RouterApi in default app APIs and AppRoot

Signed-off-by: Paul Schultz <pschultz@pobox.com>
Made-with: Cursor

schultzp2020 committed:
refactor: migrate plugins and packages to RouterApi

Migrate all plugin and package consumers away from direct
react-router-dom imports to the new RouterApi abstraction.

Key changes:
- Replace react-router-dom imports with RouterApi hooks/components
  across catalog, scaffolder, techdocs, search, home, kubernetes,
  auth, org, and other plugins
- Migrate Storybook stories to TestMemoryRouterProvider
- Remove react-router-dom peer dependencies from migrated packages
- Update docs-ui components to use RouterApi

Signed-off-by: Paul Schultz <pschultz@pobox.com>
Made-with: Cursor

schultzp2020 committed:
refactor: decouple BUIProvider from react-router-dom via versioned context bridge

Introduce a BUIRouterProvider that passes useNavigate and useHref through
a versioned context, removing BUIProvider's direct dependency on
react-router-dom. This allows @backstage/ui to remain router-agnostic
while still integrating with react-aria RouterProvider when a router
context is available.

Made-with: Cursor
Signed-off-by: Paul Schultz <pschultz@pobox.com>
Made-with: Cursor

schultzp2020 commented:
replaced by https://github.com/backstage/backstage/pull/33900
